src/topology/builder.rs

```diff
@@ -218,41 +219,53 @@ pub async fn build_pieces(
             tracing::Span::none(),
         )
         .unwrap();
-        let mut input_rx = crate::utilization::wrap(Pin::new(input_rx));
 
         let task = match transform {
             Transform::Function(mut t) => {
-                let (output, control) = Fanout::new();
+                let (mut output, control) = Fanout::new();
 
-                let transform = input_rx
+                let mut input_rx = input_rx
                     .filter(move |event| ready(filter_event_type(event, input_type)))
-                    .ready_chunks(128) // 128 is an arbitrary, smallish constant
-                    .inspect(|events| {
+                    .ready_chunks(128); // 128 is an arbitrary, smallish constant
+
+                let mut timer = crate::utilization::Timer::new();
+                let mut last_report = Instant::now();
+
+                let transform = async move {
+                    timer.start_wait();
+                    while let Some(events) = input_rx.next().await {
+                        timer.stop_wait();
+                        if last_report.elapsed().as_secs() >= 5 {
+                            timer.report();
+                            last_report = Instant::now();
```

Comment:
All of these lines separately fetch `Instant::now()` at some level (`elapsed` calls it internally), which is _cheap_ but not _free_. Is there a reasonable way to have, say, `stop_wait` return the `Instant` so the others can reuse it?

Also, this combination of timer + last_report appears to be a common pattern. Do all timer/utilization users do this? Could it be moved into shared code in `src/utilization.rs`?

Reply:
> All of these lines separately fetch `Instant::now()` at some level (`elapsed` calls it internally), which is _cheap_ but not _free_. Is there a reasonable way to have, say, `stop_wait` return the `Instant` so the others can reuse it?

Good idea, I pushed a commit that does most of this.

> Also, this combination of timer + last_report appears to be a common pattern. Do all timer/utilization users do this? Could it be moved into shared code in `src/utilization.rs`?

This one I'm going to defer since I'm planning to rework the whole reporting pattern in an upcoming PR. Definitely agree it'd be good to centralize.